proxy/coco/src/peer/run_state.rs

```diff
@@ -394,61 +424,74 @@ impl RunState {
-                self.status = Status::Online {
-                    connected: self.connected_peers.len(),
-                };
+                match state {
+                    Status::Offline => {
+                        self.status = Status::Online {
+                            connected: self.connected_peers.len(),
+                        };
 
-                vec![]
+                        vec![]
+                    },
+                    Status::Started => {
+                        // Sync with first incoming peer.
+                        //
+                        // In case the peer is configured to sync on startup we start syncing,
+                        // otherwise we go online straight away.
+                        // TODO(xla): Also issue sync if we come online after a certain period of
+                        // being disconnected from any peer.
+                        if self.config.sync.on_startup {
+                            self.status = Status::Syncing {
+                                synced: 0,
+                                syncs: 0,
+                            };
+                            self.status_since = Instant::now();
+
+                            vec![
+                                Command::SyncPeer(peer_id),
+                                Command::StartSyncTimeout(self.config.sync.period),
+                            ]
+                        } else {
+                            self.status = Status::Online {
+                                connected: self.connected_peers.len(),
+                            };
+                            self.status_since = Instant::now();
+
+                            vec![]
+                        }
+                    },
+                    // Issue syncs until we reach maximum amount of peers to sync with.
+                    Status::Syncing { syncs, .. } if *syncs < self.config.sync.max_peers => {
+                        vec![Command::SyncPeer(peer_id)]
+                    },
+                    // Update status with its connected peers.
+                    Status::Online { .. } => {
+                        self.status = Status::Online {
+                            connected: self.connected_peers.len(),
+                        };
+                        vec![]
+                    },
+                    // Noop
+                    Status::Stopped | Status::Syncing { .. } => vec![],
+                }
             },
             // Remove peer that just disconnected.
             (_, ProtocolEvent::Disconnecting(peer_id)) => {
-                self.connected_peers.remove(&peer_id);
+                if let Some(counter) = self.connected_peers.get_mut(&peer_id) {
+                    *counter -= 1;
+
+                    if *counter == 0 {
+                        self.connected_peers.remove(&peer_id);
+                    }
+                } else {
+                    unreachable!();
```

Comment:
Why do we need this `else` branch?

Reply:
This branch should never been hit, hence the unreachable. Haven't found a way to encode this invariance.

Reply:
But that branch doesn't need to be included. So could we log an error instead or something? Like:

```rust
log::error!("The impossible has happened, somehow we disconnected from '{}' without already being connected to them", peer_id);
```